<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        
        
        
        <link rel="shortcut icon" href="../img/favicon.ico">

	<title>Usage - Plasmid Extractor</title>

        <link href="../css/bootstrap-custom.min.css" rel="stylesheet">
        <link href="../css/font-awesome-4.0.3.css" rel="stylesheet">
        <link rel="stylesheet" href="../css/highlight.css">
        <link href="../css/base.css" rel="stylesheet">

        <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
        <!--[if lt IE 9]>
            <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
            <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
        <![endif]-->

        
    </head>

    <body>

        <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>

            <!-- Main title -->
            <a class="navbar-brand" href="..">Plasmid Extractor</a>
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
            <!-- Main navigation -->
            <ul class="nav navbar-nav">
            
            
                <li >
                    <a href="..">Home</a>
                </li>
            
            
            
                <li >
                    <a href="../installation/">Installation</a>
                </li>
            
            
            
                <li class="active">
                    <a href="./">Usage</a>
                </li>
            
            
            
                <li >
                    <a href="../customization/">Customizing Plasmid Extractor</a>
                </li>
            
            
            
                <li >
                    <a href="../FAQ/">FAQ</a>
                </li>
            
            
            </ul>

            <!-- Search, Navigation and Repo links -->
            <ul class="nav navbar-nav navbar-right">
                <li >
                    <a rel="next" href="../installation/">
                        <i class="fa fa-arrow-left"></i> Previous
                    </a>
                </li>
                <li >
                    <a rel="prev" href="../customization/">
                        Next <i class="fa fa-arrow-right"></i>
                    </a>
                </li>
                
                <li>
                    <a href="https://github.com/lowandrew/Plasmid_Assembler">
                        
                            <i class="fa fa-github"></i>
                        
                        GitHub
                    </a>
                </li>
                
            </ul>
        </div>
    </div>
</div>

        <div class="container">
            <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
    
        <li class="main active"><a href="#usage-with-a-pip-install">Usage With a Pip Install</a></li>
        
    
        <li class="main "><a href="#usage-with-docker">Usage with Docker</a></li>
        
    
        <li class="main "><a href="#a-few-usage-examples">A Few Usage Examples</a></li>
        
            <li><a href="#mandatory-arguments">Mandatory Arguments</a></li>
        
            <li><a href="#optional-arguments">Optional Arguments</a></li>
        
    
    </ul>
</div></div>
            <div class="col-md-9" role="main">

<h1 id="usage-with-a-pip-install">Usage With a Pip Install</h1>
<p>The basic usage of Plasmid Extractor is fairly simple. As input, you will need to provide:</p>
<ul>
<li>a folder containing paired-end or unpaired reads that you think may have plasmids in them. It is assumed that forward reads contain '_R1' in their name and reverse reads contain '_R2' - see detailed usage for information on how to change this if your reads are named differently. </li>
<li>a plasmid database (included with the PlasmidExtractor distribution - see <a href="../installation/">Installation</a>) </li>
<li>resistance/virulence/incompatibility databases (also included with PlasmidExtractor distribution)</li>
<li>the path to a folder where you would like your output to be place (this folder will be created if it does not exist).</li>
</ul>
<p>For example, in order to analyze the reads in the directory <code>/home/user/reads</code> and place the output into <code>output</code> using the default plasmid database, the command would be:
<code>PlasmidExtractor.py -i /home/user/reads -o output -p databases/plasmid_db.fasta -d databases</code></p>
<p>Within the output directory, you will find the following:</p>
<ul>
<li>a CSV file called plasmidReport.csv, which shows the plasmids found for each sample, along with a score that shows (approximately) the percent identity to a reference plasmid</li>
<li>a folder for each sample, which contains a FASTA file for each plasmid found</li>
<li>a log file for each sample showing the output from each step of the pipeline </li>
<li>CSV files describing virulence, AMR, and incompatibility genes found for each sample.</li>
</ul>
<h1 id="usage-with-docker">Usage with Docker</h1>
<p>Coming soon...</p>
<h1 id="a-few-usage-examples">A Few Usage Examples</h1>
<p>The input parameters of Plasmid Extractor can be customized to your liking. Here are a few examples of different things that could be done:</p>
<ul>
<li>
<p>Set identity cutoff to 0.8, to increase sensitivity at the cost of specificity.
<code>PlasmidExtractor.py -i /home/user/reads -o output -p databases/plasmid_db.fasta -d databases -c 0.8</code></p>
</li>
<li>
<p>Don't generate sequences for plasmids found, for speedy analysis. Means you won't get typing info.
<code>PlasmidExtractor.py -i /home/user/reads -o output -p databases/plasmid_db.fasta -d databases -nc</code></p>
</li>
<li>
<p>Run in low memory mode. 
<code>PlasmidExtractor.py -i /home/user/reads -o output -p databases/plasmid_db.fasta -d databases -l</code></p>
</li>
<li>
<p>Combine low memory mode with no consensus generation for maximum speed.
<code>PlasmidExtractor.py -i /home/user/reads -o output -p databases/plasmid_db.fasta -d databases -nc -l</code></p>
</li>
</ul>
<h4 id="mandatory-arguments">Mandatory Arguments</h4>
<ul>
<li><code>-o, --output_directory</code>: The location for your output folder. Can be anything you want it to. Output folder will be created if it does not exist.</li>
<li><code>-p, --plasmid_database</code>: The path to your sequence database. This database should be an uncompressed multi-fasta file. You can create your own custom database
of plasmid sequences, or append any plasmid sequences you want to the supplied database.</li>
<li><code>-i, --input_directory</code>: The path to your input directory, which contains your paired-end reads to be analyzed. These reads can be uncompressed, or gzip/bzip2 compressed. It is assumed that forward
 reads contain 'R1' and reverse reads contain 'R2'. This assumption can be changed with the <code>-fid</code> and <code>-rid</code> options.</li>
<li><code>-d, --databases</code>: The path to databases for virulence/AMR/incompatibility gene detection.</li>
</ul>
<h4 id="optional-arguments">Optional Arguments</h4>
<ul>
<li><code>-t, --threads</code>: Number of CPUs to run your analysis on. By default this is the number of cores on your machine. Every step of the pipeline except for consensus sequence generation parallelizes quite well, so it's recommended that this be left at the default unless you need to run other jobs at the same time.</li>
<li><code>-c, --cutoff</code>: The score cutoff for a plasmid to be considered present. Corresponds roughly to the percent identity between the generated plasmid and the reference plasmid. By default, set to 0.98 to only allow plasmids to be found that are quite similar. Lower values will generally return more plasmids, with less specificity. </li>
<li><code>-r, --report</code>: Name for report that will be created in <code>--output_dir</code>. Defaults to plasmidReport.csv, but can be changed to whatever you like.</li>
<li><code>-fid, --forward_id</code>: Identifier for forward reads. Defaults to <code>_R1</code>, but if your forward reads use a different naming scheme like <code>_1</code>, specify <code>-fid _1</code> to have these recognized as forward reads.</li>
<li><code>-rid, --reverse_id</code>: Same as <code>-fid</code>, but for reverse reads. Default <code>_R2</code>.</li>
<li><code>-nc, --no_consensus</code>: Finding consensus sequences takes a fair chunk of time. If you want to skip this step and only identify the plasmids present in your sample, add this option. Adding this option means that any post-analysis of your plasmids that would usually take place (AMR detection, etc) will not occur.</li>
<li><code>-l, --low_memory</code>: Enabling this flag will make the PlasmidExtractor pipeline use substantially less memory (~7GB of RAM at peak usage instead of &gt;20GB). May cause slight drops in sensitivity, but nothing too drastic. </li>
</ul></div>
        </div>

        <footer class="col-md-12">
            <hr>
            
            <center>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</center>
        </footer>

        <script src="../js/jquery-1.10.2.min.js"></script>
        <script src="../js/bootstrap-3.0.3.min.js"></script>
        <script src="../js/highlight.pack.js"></script>
        <script src="../js/base.js"></script>
    </body>
</html>