<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Installation - Plasmid Extractor</title>
  

  <link rel="shortcut icon" href="../img/favicon.ico">

  
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="../css/highlight.css">

  
  <script>
    // Current page data
    var mkdocs_page_name = "Installation";
  </script>
  
  <script src="../js/jquery-2.1.1.min.js"></script>
  <script src="../js/modernizr-2.8.3.min.js"></script>
  <script type="text/javascript" src="../js/highlight.pack.js"></script>
  <script src="../js/theme.js"></script> 

  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> Plasmid Extractor</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        <ul class="current">
          
            <li>
    <li class="toctree-l1 ">
        <a class="" href="..">Home</a>
        
    </li>
<li>
          
            <li>
    <li class="toctree-l1 current">
        <a class="current" href="./">Installation</a>
        
            <ul>
            
                <li class="toctree-l3"><a href="#system-requirements">System Requirements</a></li>
                
                    <li><a class="toctree-l4" href="#installation-using-docker">Installation Using Docker</a></li>
                
                    <li><a class="toctree-l4" href="#installing-using-pip">Installing Using Pip</a></li>
                
            
            </ul>
        
    </li>
<li>
          
            <li>
    <li class="toctree-l1 ">
        <a class="" href="../usage/">Usage</a>
        
    </li>
<li>
          
            <li>
    <li class="toctree-l1 ">
        <a class="" href="../FAQ/">FAQ</a>
        
    </li>
<li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">Plasmid Extractor</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>Installation</li>
    <li class="wy-breadcrumbs-aside">
      
        
          <a href="https://github.com/lowandrew/Plasmid_Assembler" class="icon icon-github"> Edit on GitHub</a>
        
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="system-requirements">System Requirements</h1>
<p>PlasmidExtractor has been developed and tested on Linux-based systems (more specifically Ubuntu and Mint). Any Linux/Unix-based system (including MacOS) should be able to run the pipeline, while users of Windows-based systems will have to use Docker in order to have the pipeline run.</p>
<p>In terms of system specs the more CPUs your system has, the better, as almost every step of the pipeline is multi-threaded. RAM requirements can be somewhat heavy, with some parts of the pipeline more than 20 GB due to the size of the database that PlasmidExtractor uses. 32 GB or RAM is recommended. If your machine does not meet these requirements, PlasmidExtractor can be run in a low-memory mode, where memory usage should peak at ~7GB, enabling the pipeline to run on a much wider range of machines.</p>
<h2 id="installation-using-docker">Installation Using Docker</h2>
<p>As PlasmidExtractor has a fair number of dependencies, the easiest way to get it installed and working on your machine is using docker. Instructions on docker installation can be found <a href="https://docs.docker.com/engine/installation/">here</a>.</p>
<p>Once you have docker installed, you can load the PlasmidExtractor image by booting a terminal and typing:</p>
<p><code>docker pull olcbioinformatics/cowpig</code>.</p>
<p>The pipeline can then be run with: </p>
<p><code>docker run -i -v /path/to/your/sequences/:/sequences olcbioinformatics/cowpig python3 /home/Extractor.py -i /sequences -o /sequences/output -sdb /home/new_database.fasta</code></p>
<p>Where <code>/path/to/your/sequences/</code> is the folder with the FASTQ files you want analyzed. A folder called output will be created in your input folder that will contain your results.</p>
<p>Currently, the docker image is unable to create the output visualizations - this will be addressed in the future.</p>
<h2 id="installing-using-pip">Installing Using Pip</h2>
<h4 id="executable">Executable</h4>
<p>PlasmidExtractor can also be installed using pip. Use of a virtual environment for PlasmidExtractor is highly recommended. To create a virtualenv:</p>
<ul>
<li>Create an empty directory (i.e. <code>mkdir ~/Virtual_Environments/PlasmidExtractor</code>)</li>
<li>Virtualenv that directory (<code>virtualenv -p /usr/bin/python3 ~/Virtual_Environments/PlasmidExtractor</code>)</li>
<li>Activate the virtualenv (<code>source ~/Virtual_Environments/PlasmidExtractor/bin/activate</code>)</li>
<li>Install PlasmidExtractor - this should also install any python packages necessary for PlasmidExtractor to run. (<code>pip install plasmidextractor</code>)</li>
</ul>
<p>With this done, you'll need to make sure that any necessary dependencies are installed.</p>
<h4 id="dependencies">Dependencies</h4>
<p>With that done, you will need to make sure that you have all of the depedencies for PlasmidExtractor installed and present on your $PATH. The dependencies for PlasmidExtractor are:</p>
<ul>
<li><a href="http://www.htslib.org/download/">samtools &gt;= 1.6</a></li>
<li><a href="http://www.htslib.org/download/">bcftools &gt;= 1.6</a></li>
<li><a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">ncbi-blast &gt;= 2.2.31</a></li>
<li><a href="http://bedtools.readthedocs.io/en/latest/content/installation.html">bedtools &gt;= 2.25.0</a></li>
<li><a href="https://jgi.doe.gov/data-and-tools/bbtools/">BBtools &gt;= 37.23</a></li>
<li><a href="https://github.com/marbl/Mash/releases">mash &gt;= 2.0</a></li>
<li><a href="http://sun.aei.polsl.pl/REFRESH/index.php?page=projects&amp;project=kmc&amp;subpage=download">kmc &gt;= 3.0</a></li>
<li><a href="https://www.python.org/downloads/">python &gt;= 3.5</a></li>
</ul>
<p>For instructions on how to add programs to your $PATH, click <a href="https://askubuntu.com/questions/60218/how-to-add-a-directory-to-the-path">here</a>.</p>
<p>If all of your dependencies are properly installed, you should now be able to run PlasmidExtractor. If installation of a dependency has not worked, you should get a <em>ModuleNotFoundError</em> for the depedency that has not been able to run properly.</p>
<p>For visualization of sourmash results, you may need to install the python3-tk library. On a debian-based system, you can do this with: <code>apt-get install python3-tk</code></p>
<h4 id="databases">Databases</h4>
<p>Once you have the executable and dependencies installed, you'll just need to download the databases that ConFindr depends on.</p>
<p>The databases necessary for PlasmidExtractor are hosted on FigShare.</p>
<p>To download them, use the following command in the folder you want to download the databases to:</p>
<p><code>wget https://ndownloader.figshare.com/files/9827323  &amp;&amp; tar xf 9827323</code></p>
<p>This will create a folder in you current working directory called databases.</p>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../usage/" class="btn btn-neutral float-right" title="Usage"/>Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href=".." class="btn btn-neutral" title="Home"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
	  
        </div>
      </div>

    </section>

  </div>

<div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a class="icon icon-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href=".." style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../usage/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>

</body>
</html>
